### ISSUE TITLE
Akeneo should follow Semantic Versioning.

### ISSUE THREAD

**Issue by squaresurf** [akeneo/pim-community-dev#3261]
It would be so tremendously helpful if this project did not introduce any backwards compatibility breaks within a minor or patch version. It is so extremely frustrating to see that the project "follows" the [semantic versioning](http://semver.org/) format to find out that it doesn't actually conform to [SemVer](http://semver.org/). The frustration specifically stems from the experience you get anytime you want to upgrade in order to get bug fixes and then must look through all your code and fix any BC breaks.

[SemVer](http://semver.org/) would solve this sort of an issue and take Akeneo to the next level in terms of community respect.

**Comment by nidup** [comment]
Thank you for your kind and clear message, it's a good opportunity for us to communicate on this topic.

First, we easily understand your concern and we would love to respect the SemVer!

We release minor version every 6 months which should be from SemVer point of view, a major version.

You're totally right on this point.

But releasing a major version every 6 months is really difficult from business/marketing point of view, mainly because of the meaning it would have for our customers.

Moreover the product is still young and continues to evolve a lot on both functional and technical sides, it's really difficult for us to freeze the internal API right now and we're convinced that if we do it, we'll not be able to improve our quality standard and to release valuable features.

So, even if we don't respect the SemVer, we have the following rule:
 - we never introduce BC Breaks in patch versions
 - we introduces BC Breaks in minor versions but we provide scripts and documentation (upgrade.md) to support the migration (WIP for the future 1.4).

For minor versions, we strongly advise to plan the migrations, in the other hand, the patches can be directly applied without special actions.

We'll re-discuss this point in the team for sure, thank you again for your feedback!

**Comment by squaresurf** [comment]
Thank you for explaining your current process. However my biggest frustration has been the BC Breaks between patch versions. As you can see in [CHANGELOG-1.3.md](https://github.com/akeneo/pim-community-dev/blob/master/CHANGELOG-1.3.md) there were BC Breaks between 1.3.5 and 1.3.6. It would be a tremendous help if you would not introduce BC Breaks in patch versions so that we can all benefit from bug fixes with minimal effort.

It's also unfortunate that a business/marketing point of view is affecting the respectability of the engineering quality of Akeneo. In my **opinion**, the versioning scheme should not be under the control of the marketers.

**Comment by nidup** [comment]
We do our best to not introduce BC Break in patches but it already happened in the past.

Don't hesitate to ping us if you encounter any issue with a BC break when applying a patch, this feedback is very valuable to push us to enhance our practices and the perceived quality.

Thank you.

**Comment by squaresurf** [comment]
Thank you for all your hard work on Akeneo.

**Comment by sagikazarmark** [comment]
I mainly agree with @squaresurf, but I also understand marketing reasons as well. I think that in cases like this it is enough to say "do your best not to break BC, but do it if it is absolutely necessary". Keeping this in mind very strictly make developers think more about a specific problem which takes more time, but ensures that the API is not randomly broken and there is no viable solution for that problem. But it is hard to follow this concept as "doing the best" can mean different things to different peoples.

So my concern is that it is not communicated enough that BC breaks are allowed, but they should be avoided.

I've read somewhere that it is a possible solution to introduce punishments for BC breaks. For example: buy a beer for each colleague per BC break. :stuck_out_tongue_winking_eye: Not sure if it is possible in this case.

**Comment by squaresurf** [comment]
I think the biggest thing is that the `#.#.#` version format indicates SemVer. It would help tremendously if they didn't use that at all since they don't follow SemVer. That being said, I'm not sure they could use another format and make the project available via Composer.

**Comment by sagikazarmark** [comment]
> I think the biggest thing is that the #.#.# version format hints at SemVer.

To be completely fair: it might hint at SemVer as it is the most popular versioning strategy (It's not a coincidence!), but it is not the only one and this version pattern is quite common (at least the #.# form is). However I think SemVer is rather about communication than the actual numbering: "We are using semver, so you can expect preserved BC in minor and patch releases...".

**Comment by nidup** [comment]
Thank you for your feedback.

Once again, we forbid all BC Breaks in patches, it can occur exceptionally and, if we have to do, we discuss with the team, it's a choice, not a mistake.

About the definition of Backward Compatibility itself, we really like the approach of Symfony and we would go to this direction for Akeneo PIM http://symfony.com/doc/current/contributing/code/bc.html (ie defining api and internal classes).

As you can see, even a so popular and mature framework cannot afford a very strict respect of BC on all classes and all public/protected methods of the stack in a minor release.

**Comment by sagikazarmark** [comment]
Are there any references to this in the documentation?

**Comment by willy-ahva** [comment]
Hi !

There is no description of how we manage to increment a version and how we manage BC breaks or not in Major, Minor or Patch version. I understand well it can be very frustrating to discover at your expense how we work with. We currently hard working on the documentation and all proposition are very welcome.
I added a ticket about it in our Idea Manager :)

Thank you very much for your feedback !
Willy

**Comment by a2xchip** [comment]
I think when you use application in production it is must to test it on development/testing machine before upgrading on production. But anyway semantic versioning is community standard today and it would be nice if Akeneo PIM will respect it too in the feature.
